### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by diego3g** [diego3g/umbriel#118]
I sent an email for 18k contacts and noticed some CPU and memory problems.

## Summary

- The HTTP service got a high CPU usage due to AWS SNS notifications;
- The queue service got a high CPU usage due to queue processing to send emails;
- The HTTP service got a high memory usage when adding the batch of jobs to the queue;

## Errors

When receiving some AWS events, the HTTP service was logging some errors with this trace:
```
PrismaClientKnownRequestError2 [PrismaClientKnownRequestError]: Unique constraint failed on the fields: (`message_id`,`contact_id`)
    at cb (/home/node/app/node_modules/@prisma/client/runtime/index.js:34800:17)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async PrismaRecipientsRepository.saveWithEvents (/home/node/app/dist/modules/broadcasting/repositories/prisma/PrismaRecipientsRepository.js:41:5)
    at async RegisterEvent.execute (/home/node/app/dist/modules/broadcasting/useCases/RegisterEvent/RegisterEvent.js:60:5)
    at async RegisterEventController.handle (/home/node/app/dist/modules/broadcasting/useCases/RegisterEvent/RegisterEventController.js:90:22)
    at async /home/node/app/dist/infra/http/adapters/ExpressRouteAdapter.js:15:26 {
  code: 'P2002',
  clientVersion: '2.28.0',
  meta: { target: [ 'message_id', 'contact_id' ] }
} 
```

### Infrastructure

- [ ] Split the HTTP service that receives SNS notifications in a new service with (100mCPU, 144mb RAM and 35% CPU autoscale threshold);
- [x] Increased the amount of CPU of Queue service to 400m CPU; 

### Application

- [ ] todo

**Comment by diego3g** [comment]
Did some benchmarking around queue job inserting using BullMQ and got the following results:

- Inserting 20k jobs individually on the queue leads to 13mb RAM usage around 0.055ms;
- Inserting 20k jobs in a single batch on the queue leads to 190mb RAM usage around 956.88ms;

Why inserting in batch is more expensive than inserting one by one? I don't know, but it is.

My guesses about the memory usage being around the creation of a big recipients array in the JavaScript side were wrong, the problem was inserting to queue.

**Comment by diego3g** [comment]
Did some benchmarks with Bee Queue, but it didn't result in a better performance, actually it changed nothing.

**Comment by diego3g** [comment]
Removed two queries when receiving notifications from Amazon SNS.

Now we don't need to check for recipient existence nor create the events separatelly from upserting the recipient.

**Comment by diego3g** [comment]
Split the SNS webhook HTTP service on c81033171578d5859383b897fed4b41007bbff76.

**Comment by pellizzetti** [comment]
@diego3g 

> Inserting 20k jobs individually on the queue leads to 13mb RAM usage around 430ms (This uses high CPU, but maybe it's better to have higher CPU than memory usage);

Touchy subject, memory is usually considerably cheaper than vCPU[0]. Some numbers around the CPU usage in both tests would be nice!

[0]
|  Item  | price |
|:------:|:---------------------:|
| vCPU   | $0.033174 / vCPU hour |
| Memory | $0.004446 / GB hour   |

It's also easier to manage/scale/allocate memory than CPU.

Besides all that, applications in Node.js tend to consume more CPU than memory. Since we [currently] work almost exclusively with this technology, depending on the usage of each resource, it may be worth to sacrifice some RAM. (:

**Comment by diego3g** [comment]
Hey @pellizzetti, I tried making some CPU benchmarking but I couldn't manage to get the actual results of CPU consumption. Do you have any tips for me on that?

**Comment by pellizzetti** [comment]
@diego3g 

If you created new, isolated scripts for your benchmarking, you won't need to handle startup/setup time, so, I guess a simple `time` would do it, e.g `time node <script>`. Just a quick reminder, it would be nice to run it some times to get the avg CPU usage of each script.

**Comment by pellizzetti** [comment]
If you want to take a deeper dive, take some time to look over `perf`/`perf_events`[0].

[0] https://www.brendangregg.com/perf.html

**Comment by Deividy** [comment]
> Why inserting in batch is more expensive than inserting one by one? I don't know, but it is.

well, it all comes down to how the `.addBulk` is implemented, if you take a close look at [that method](https://github.com/taskforcesh/bullmq/blob/18b0b7954313274a61fcc058380bfb9d682c059d/src/classes/flow-producer.ts#L167-L179) you will see that there is no bulk going on, and the actual bulk process in `BullMQ` is [adding the job and doing some child home work](https://github.com/taskforcesh/bullmq/blob/18b0b7954313274a61fcc058380bfb9d682c059d/src/classes/flow-producer.ts#L192-L253)

that's why it's more expensive.

(also, there is no bulk/mass insert support in redis)

now, to measure the *exactly* CPU used by a Node.js process you can try the [process.getCpuUsage()](https://nodejs.org/api/process.html#process_process_cpuusage_previousvalue) before and compare after the process, [the implementation](https://github.com/nodejs/node/blob/e83c8aef4b787b531f23ad7abb498fe81db5ee83/src/node_process_methods.cc#L109-L130) uses the libuv uv_getrusage which is really the [kernel getcpu](https://man7.org/linux/man-pages/man2/getcpu.2.html)

measure the exact CPU usage is a tricky task;
but, I think that the question here is; why? why are you inserting 20k jobs in a redis queue anyway? this sounds wrong.

**Comment by Deividy** [comment]
wow, I thought this would be a fun quest for a friday night, jumped in and I got your answer :)

> Why inserting in batch is more expensive than inserting one by one? I don't know, but it is.

> tldr; 
> because extraneous loops allocating different objects and a bunch of strings before sending to redis; when inserting only one you don't allocate all these objects/strings, and you can save some resources with fewer interactions in these objects.

here is the full story:

- calling `addBulk()` in BulkMQ triggers *multiple* loops in the whole data of objects passed, initializing a lot of new Objects.
testing on my Macbook with simple 20k objects ({ foo: 'bar' }), **only** [this loop](https://github.com/taskforcesh/bullmq/blob/3b9b79dbdd754ab66c3948e7e16380f2d5513262/src/classes/job.ts#L191-L205) takes approximately 150ms~ of pure CPU

- the thing got bad to memory when we call the line #207
```typescript
const result = (await multi.exec()) as [null | Error, string][];
```

this basically loops for all the 20k objects that we just added, and allocate a new array of strings, around here:
- https://github.com/luin/ioredis/blob/e5615da8786956df08a9b33b6cd4dd31e6eaa759/lib/pipeline.ts#L321-L345

at this point, the `scripts[]` is an array with 20k strings [like this](https://github.com/OptimalBits/bull/blob/6345c79db58fc15a60d5f46c820222f979922b4c/lib/commands/addJob-6.lua)

so, there is a bunch of loops in all data:

- first the `addBulk()` itself contains a `.map()`  (which will loop all 20k objects)
- https://github.com/taskforcesh/bullmq/blob/afdaac6b072c7af5973222cc7fb69f3f138f3b0b/src/classes/queue.ts#L211-L219
- then, the `createBulk()` again calls a map in all the 20k objects: 
- https://github.com/taskforcesh/bullmq/blob/3b9b79dbdd754ab66c3948e7e16380f2d5513262/src/classes/job.ts#L191-L194
- after the map the `createBulk()` does another loop in all the objects, this time adding the jobs to the redis cliente queue:
- https://github.com/taskforcesh/bullmq/blob/3b9b79dbdd754ab66c3948e7e16380f2d5513262/src/classes/job.ts#L198-L205
- the `addJob()` itself is a little expensive if your objects are big since it calls a `JSON.stringify()` in every one: 
- https://github.com/taskforcesh/bullmq/blob/3b9b79dbdd754ab66c3948e7e16380f2d5513262/src/classes/job.ts#L893
- the most usage of memory is due by `ioredis` allocating the 20k rows as strings (my chrome exploded when I was debugging there :p)
- https://github.com/luin/ioredis/blob/e5615da8786956df08a9b33b6cd4dd31e6eaa759/lib/pipeline.ts#L321-L345
- and, if you are not in cluster mode, the `ioredis` will send the 20k strings at once to process: 
- https://github.com/luin/ioredis/blob/e5615da8786956df08a9b33b6cd4dd31e6eaa759/lib/pipeline.ts#L364

---

performance issues are tricky; you might think that using `time` will get the correct time used by the process or that using the [cpuUsage()](https://nodejs.org/api/process.html#processcpuusagepreviousvalue) will get the right answers, but most part of the time is not that simple, you will have other processes in the machine, other tasks running in your node application and so on.

to understand this kind of issue you must find first where is your issue, CPU, or memory (in this case, both), then go dig in the code, read and understand what folks are doing, `debugger` and `console.log` are your best friends here.

✌️

**Comment by Deividy** [comment]
thought this would be a fun quest, decided to jump in and I got your answer :)

> Why inserting in batch is more expensive than inserting one by one? I don't know, but it is.

> tldr; 
> because extraneous loops allocating different objects and a lot of strings (JSON.stringify()) before sending to redis; when inserting only one you don't allocate all these objects, and you can save some resources with fewer interactions.

here is the full story:

- calling `addBulk()` in BullMQ triggers *multiple* loops in the whole data, initializing a lot of new Objects.
testing on my Macbook with simple 20k objects ({ foo: 'bar' }), **only** [this loop](https://github.com/taskforcesh/bullmq/blob/3b9b79dbdd754ab66c3948e7e16380f2d5513262/src/classes/job.ts#L191-L205) takes approximately 150ms~ of pure CPU

- and it got worse in line #207
```typescript
const result = (await multi.exec()) as [null | Error, string][];
```

this basically loops for all the 20k objects that we just added, checking some flags and building a `scripts[]` array
- https://github.com/luin/ioredis/blob/e5615da8786956df08a9b33b6cd4dd31e6eaa759/lib/pipeline.ts#L321-L345


then it loops **again** all the 20k objects and sends one by one to redis 
- https://github.com/luin/ioredis/blob/e5615da8786956df08a9b33b6cd4dd31e6eaa759/lib/pipeline.ts#L444-L446

so, there is a bunch of loops in all data:

- first the `addBulk()` itself contains a `.map()`  (which will loop all 20k objects)
- https://github.com/taskforcesh/bullmq/blob/afdaac6b072c7af5973222cc7fb69f3f138f3b0b/src/classes/queue.ts#L211-L219
- then, the `createBulk()` again calls a map in all the 20k objects: 
- https://github.com/taskforcesh/bullmq/blob/3b9b79dbdd754ab66c3948e7e16380f2d5513262/src/classes/job.ts#L191-L194
- after the map the `createBulk()` does another loop in all the objects, this time adding the jobs to the redis cliente queue:
- https://github.com/taskforcesh/bullmq/blob/3b9b79dbdd754ab66c3948e7e16380f2d5513262/src/classes/job.ts#L198-L205
- the `addJob()` itself is a little expensive if your objects are big since it calls a `JSON.stringify()`: 
- https://github.com/taskforcesh/bullmq/blob/3b9b79dbdd754ab66c3948e7e16380f2d5513262/src/classes/job.ts#L893
- then the `ioredis` just loops again through all data to build their `scripts[]`
- https://github.com/luin/ioredis/blob/e5615da8786956df08a9b33b6cd4dd31e6eaa759/lib/pipeline.ts#L321-L345
- and loops again through all data to send each command, one by one to redis
- https://github.com/luin/ioredis/blob/e5615da8786956df08a9b33b6cd4dd31e6eaa759/lib/pipeline.ts#L444-L446

in the end, you have a bunch of copies of the original array; every `.map()` allocates a new array, BullMQ maintains the full queue, and ioredis also maintains the full queue with all objects.
that's why this `addBulk()` is memory expensive, and it's CPU expensive because it has *a lot* of loops in all data and every loop blocks the event loop and CPU.

---

performance issues are tricky; you might think that using `time` will get the correct time used by the process or that using the [cpuUsage()](https://nodejs.org/api/process.html#processcpuusagepreviousvalue) will get the right answers, but most part of the time is not that simple, you will have other processes in the machine, other tasks running in your node application and so on.

to understand this kind of issue you must dig into the code, read and understand what folks are doing, `debugger` and `console.log` are your best friends here, external tools will only point the symptom, rarely will show you the issue. sometimes it's better to actually read the code and fully understand what's going on than benchmarking different modules, they might be using the same strategy.

a base ground rule when working with big data in javascript;
- every loop blocks the CPU/Eventloop for `N ms` (depends on the complexity of loop and array size)
- every `.map/reduce/split/JSON.stringify/anything that allocates a new object()` in this data will use more memory, sometimes it will double the usage or even go to 3x+, be careful and work reducing the iterations with this data.


✌️
